### DISCUSSION TITLE
How to Implement Middleware in Remix?

### DISCUSSION THREAD

**Discussion by goldcoders** [remix-run/remix#1432] [Q&A] [answered]
Its a bit sad if i need to add logic on each page to handle a page redirect if user is auth, or have proper permission.
Is there a way we can declare on global scope a middleware?

I came from laravel background , so im bit confuse on implementation on React / Node JS.

Hope someone can share a bit of snippet or packages that can handle middleware for remix :) 

thanks

**Comment by sergiodxa** [comment, marked as answer]
Instead of middleware, you can call a function directly inside the loader, this will also be more explicit. If you want to early return a response from those "middlewares" Remix let you throw the response object.

For example, if you wanted to check the user has a certain role you could create this function:

```ts
async function verifyUserRole(request: Request, expectedRole: string) {
  let user = await getAuthenticatedUser(request); // somehow get the user
  if (user.role === expectedRole) return user;
  throw json({ message: "Forbidden" }, { status: 403 });
}
```

And in any loader call it this way:

```ts
let loader: LoaderFunction = async ({ request }) => {
  let user = await verifyUserRole(request, "admin");
  // code here will only run if user is an admin
  // and you'll also get the user object at the same time
};
```

Another example could be to require HTTPS

```ts
function requireHTTPS(request: Request) {
  let url = new URL(request.url);
  if (url.protocol === "https:") return;
  url.protocol = "https:";
  throw redirect(url.toString());
}

let loader: LoaderFunction = async ({ request }) => {
  await requireHTTPS(request);
  // run your loader (or action) code here
};
```

**Reply by sergiodxa** [reply]
This is also better for TS, with middlewares a request could have anything inside it, with this you keep the request clean.

**Reply by cloudcompute** [reply]
What if I have 100 loaders.. calling verifyUserRole() and probably some other global functions like logging in every loader is painful.

**Reply by chikamichi** [reply]
It’s just code, you may hide function-chaining (aka. middlewares chain) behind an internal API:

``` js
import myMiddlewares from "~/lib/middlewares";

let loader: LoaderFunction = async ({ request }) => {
  await myMiddlewares(request); // use any DIY- or lib-provided middleware management within myMiddlewares
}
```

or even better I guess, using a callback (giving you control over the best way/moment to call the loader function’s business logic, hence making it possible/easier to support pre- and post- middlewares, declarative chaining, etc.)

``` js
import myMiddlewares from "~/lib/middlewares";

let loader: LoaderFunction = myMiddlewares(({ request }) => { … });
```

or whatever.

And of course, you may use a custom, dev-controlled server (instead of Remix’ default server), allowing you to shift middleware management from Remix’ backend, to that server’s code.

**Reply by esamattis** [reply]
Unfortunately higher order function loader like the myMiddlewares() will break server code pruning and your server code will get loaded for browsers too. See https://remix.run/docs/en/v1/guides/constraints#higher-order-functions

**Reply by chikamichi** [reply]
Fair point. Haven’t tried fiddling around with lazy evaluation yet, I guess one should be able to overcome this limitation.
First API (calling `await myMiddlewares(request)`) can be tweaked to the same effect though: `await myPreLoaderMiddlewares(request)` + `await myPostLoaderMiddlewares(request)`, just a tad bit more boilerplate.

**Reply by akomm** [reply]
The issue is, you have to do it in all loaders and actions as well. Otherwise a user can continue hit action API directly.
Now imagine you your "active" loaders in the tree are being triggered on changes, all of them running the code. Not like you couldn't cache the db fetch and refetch only once every 10 seconds. But problem A, follows problem B, follows problem C, ... you see where it goes?

I think there will be many insecure websites done with remix, because people didn't see their actions are exposed. Because it appears to "work" when testing it from within the app. Having it in a layout route, or just the loaders isn't enough. Especially for a framework that advertises with "like old school php", one expects a guard on the parent route should cover the child routes and that the behavior SSR (renders the whole page = security works), but SPA suddenly behaves differently in regards to guarding the routes. However the taught behavior of remix is, that it doesn't matter whether you navigate client or full reload, and it can even fall back.

I don't say it does not make any sense. I just tell, its very inconvenient and unintuitive and invites to terrible security holes.

**Reply by jaschaio** [reply]
I see this very similar to akomm. Especially as for loaders all loaders are called, but for actions on the deepest nested action does. So you need to implement authentication in each action and remember to call it at the top of the action. This seems a lot more error prone that simply having a single point of entry like a middleware.

**Reply by akomm** [reply]
@jaschaio even loaders are insecure. Sure, remix on client "refetches" all the active (route) loaders and if any of them has a redirect to login it would redirect you client-side. But outside the app you can just hit the loaders directly. Yes, the redirecting one if you hit it directly would not leak any data, but the others up and down the hierarchy would.

**Reply by jaschaio** [reply]
Well perfect example then how loaders and actions code is highly repetitive and if you forget it even in a single one you have a problem. Thanks for letting me know, it's not intuitive.

**Reply by kiliman** [reply]
One thing I do (using the Express adapter) is to check for the auth cookie before handing the request to Remix. You have to explicitly whitelist "unprotected" routes, otherwise it will automatically redirect to login if cookie not present. This way all routes are protected by default. <https://github.com/kiliman/express-auth-example/blob/main/server/index.js>

With that said, I know they are considering adding pre/post request/response hooks, so a form of middleware will be possible in the future.

**Reply by akomm** [reply]
@kiliman this bears many problems. 
1. Lack of colocation to the routes, you have to sync the white/blacklist url "strings" to the structure in routes folder
2. You can not integrate it simply into remix build step, so if you want to use typescript, you need a separate build step to output the middleware code.
3. If you want to use app code (domain types, domain functions) to perform not only authentication but also authorization, you need to be very careful, because you can blow up yourself, since the middleware is built not in same context as the remix app (code).

I don't have a single app, that does not need to do 3. for example. And most of my latest use 2. The 1. is a nasty thing but by itself could be taken as the price for other good things, but 2. & 3. is a problem.

**Reply by kiliman** [reply]
I never said it was a perfect solution. But until we get some sort of middleware/pre-request hook, it enables me to handle it in once place instead of in each loader/action.

**Reply by akomm** [reply]
I never said you'v claimed that. I just think its important to ensure this understanding for all in the discussion

**Reply by cjsauer** [reply]
Just curious, have things changed in Remix in regards to middleware? I'm not a Remix user yet, but the lack of some sort of common loader/action code flow mechanism is holding me back from adoption.

**Reply by neekey** [reply]
@cjsauer dont think so and i agree with you. I'm using Remix for a very simple project, but once the backend logic gets more complicated, i can see things will get messy. The current solution seems to create a separate API backend, and leave remix as a front end for backend service...

**Reply by ZeldOcarina** [reply]
Remix can run as an Express adapter so you can do all the middleware logic in the Express portion and handle the request with Remix when it hits that portion of the Express app!

**Reply by cjsauer** [reply]
That occurred to me as well. It’s a good strategy when using Express. It does still feel like a hole in Remix given that its goal is to “run anywhere”. The middleware pattern is such a staple, it was the first thing I looked for in the docs after scanning the tutorial.

**Reply by elyobo** [reply]
This should not be marked as the answer.

Calling manually in each loader, while more explicit, also misses the point of middleware which is that it's _not_ called in each loader. Essential functionality like authentication and authorization can be easily missed if it has to be done repetitively but it's trivial to get right everywhere with a single bit of middleware.

From the rest of the discussion here it seems like remix doesn't support middleware and so the solution is to add middleware directly to your underlying server in a way that it supports, using the framework request/response objects instead.

It looks like https://github.com/remix-run/remix/discussions/7642 is the place to follow for a remix native middleware implementation.

**Reply by snimavat** [reply]
Remix doesnt have a way to create some kid of middleware that can intercept requests ! thats a big limitation, there are dozones of use cases for a middleware

Call it in every loader? thats not a solution.

**Comment by RexFactorem** [comment]
Well, this makes kind of a pain not to be able to centralize maintenance of middlewares. There's a reason why they're so popular.

**Reply by chikamichi** [reply]
See my answer above + sergiodxa’s below.

**Comment by cloudcompute** [comment]
Consider that I am using the Remix App NodeJs Express Server. As we know that Express has numerous proven, stable middleware functions. After creating an instance of express, web apps typically have these middlewares in sequence:

```
app.use(helmet());
app.use(cookieParser());                        // needed before csurf. Injects some properties on the req object
app.use(csurf);                                       //CSRF functionality
...
app.use(logger)
```

How can I achieve this? Kindly guide.

If Remix is giving me an option to deploy a Remix app on a server of my choice like AWS, NodeJS, Cloudfare Workers, etc, it should also allow me to use each and every bit of functionality that my chosen server offers.

In my case, the incoming request should be allowed to go through a series of middleware. Pl. help me, this thing is really important before I decide to use Remix.

**Reply by sergiodxa** [reply]
If you deploy Remix to a Node server you control like Express you could setup all of those middlewares

**Reply by sergiodxa** [reply]
But if you deploy to Serverless where you don’t control the server or use Remix App Server where you also don’t control the server code then you will need to move the behavior of those middlewares to each loader inside the app code

**Reply by cloudcompute** [reply]
Got it. Thanks.

**Reply by sonnguyenhong** [reply]
Excuse me, could you please explain what does "deploy Remix to a Node server you control"? As far as I know, The Remix is bundle all express server in its library so you can not use middlewares like this unless you change the library code.

**Reply by sergiodxa** [reply]
@sonnguyenhong you can use Remix with a custom Express server, there's a template here https://github.com/remix-run/remix/tree/main/templates/express, you could use it with:
```sh
npx create-remix --template remix-run/remix/templates/express
```
This way you control the Express server and Remix behaves as a middleware.

**Reply by sonnguyenhong** [reply]
Are there any solution to eject the express server from an existed remix app? I just tried to change the library code but that's not what I wanted.

**Reply by sergiodxa** [reply]
The [rmx-cli package has a script](https://github.com/kiliman/rmx-cli/tree/main#-eject-ras) to "eject" (the team doesn't really like this term).
```sh
npx rmx-cli eject-ras
```

**Reply by sonnguyenhong** [reply]
Thank you so much for supporting me. I got the solution for my problem.

**Reply by sonnguyenhong** [reply]
I have a problem with Shopify App Remix template.
I have built my app with remix template and used npx rmx-cli eject-ras to eject the express server from Remix. But after I ran this command, the start command is also changed. Before the change, when I run npm start, it is equal to shopify app dev. But after the change, it's remix dev --manual -c \"node server.js\". And this leads to the problem that it's not create a cloudflare tunnel and automatically set up environment variable for me. Could you please help me to solve this problem?

**Reply by kiliman** [reply]
The `rmx-cli eject-ras` tool was designed to eject from the default Remix App Server template. It most likely will not support other templates, especially the Shopify template. That is designed to be used by the `shopify` CLI instead of calling `node` directly.

It looks like you can integrate with any existing Remix app using the `@shopify/shopify-app-remix` package. https://www.npmjs.com/package/@shopify/shopify-app-remix

**Comment by ZeldOcarina** [comment]
Question, maybe stupid, can't you add a global middleware in the root.tsx file inside the top-level loader or action?

**Reply by kiliman** [reply]
Not really. Loaders run in parallel, so there is no *first* loader. There's nothing special about the *root* route except that it's the parent of all routes.

Also, Remix optimizes loader calls from the client, so the root loader will only be called again during revalidation.

That said, there is an RFC in React Router (the basis of Remix) for "middleware". 

https://github.com/remix-run/react-router/discussions/9564

**Reply by ZeldOcarina** [reply]
I see. Thanks a lot for the clarification! :)

**Reply by silvenon** [reply]
@kiliman thanks a lot for the explanation, I was about to suggest a layout loader as well 😅

**Comment by pburrows** [comment]
For an express app, wouldn't you just add your middleware in the server.js file where all the other middleware is already being set up (such as morgan). I haven't actually tried this yet, but it seemed like the obvious place.

**Reply by gustavopch** [reply]
If you do so, you don’t get access to Remix’s request and context.

**Reply by lapusta** [reply]
Actually you can get access to context by providing your own [getLoadContext](https://remix.run/docs/en/main/route/loader#context).
